### PR TITLE
perf(fsdp): limit actor LM-head work to the action window

### LINKED ISSUE #1858: HF/FSDP backend: pass `logits_to_keep` to avoid materializing full logits tensor
## Problem

The HF backend's [`model_wrapper.py`](https://vscode.dev/github/NovaSky-AI/SkyRL/blob/8cd5f165d81813c41d3e00b31391500b4a256d27/skyrl/backends/skyrl_train/workers/model_wrapper.py) computes logits for **all** sequence positions, then discards observation-position logits immediately after.
```python
action_log_probs = log_probs[:, -num_actions - 1 : -1]
```
The same is done for entropies in [`PolicyWorkerBase`.](https://github.com/NovaSky-AI/SkyRL/blob/8cd5f165d81813c41d3e00b31391500b4a256d27/skyrl/backends/skyrl_train/workers/worker.py) 

For models with large vocabularies (e.g., Qwen 3.5/3.6 with vocab size 248K), this materializes a massive tensor at the `lm_head`. At bf16, 262K tokens × 248K vocab = ~123 GiB.

Often, there are significantly fewer action tokens than observation tokens, so not materializing the observation token logits can produce significant savings.  

## Proposed fix

HuggingFace transformers models already support a [`logits_to_keep`](https://github.com/huggingface/transformers/blob/main/src/transformers/models/qwen3_5/modeling_qwen3_5.py#L1620-L1632) parameter that slices `hidden_states` **before** the `lm_head` projection:

```python
slice_indices = slice(-logits_to_keep, None) if isinstance(logits_to_keep, int) else logits_to_keep
logits = self.lm_head(hidden_states[:, slice_indices, :])
```

Since `num_actions` is already available in `forward()`, passing `logits_to_keep=num_actions + 1` would avoid computing logits for observation positions entirely. No custom kernels required.

**Standard path (no sample packing):** one-line change — pass `logits_to_keep=num_actions + 1` as an int.

**With `remove_microbatch_padding` / sample packing:** action positions are interleaved across packed sequences, so `logits_to_keep` needs the tensor form (indices of action positions).

## Context

The Megatron backend already solves this via fused linear cross-entropy (#1841, #1765). This issue is about the HF/FSDP backend, which has no equivalent optimization.

### PR DESCRIPTION
**Pull request by NenoL2001**
## Summary

- Pass `logits_to_keep=num_actions + 1` on the standard text-only HF/FSDP actor path, avoiding LM-head projection over prompt positions that are discarded immediately afterward.
- Support padding-free packed batches with a tensor-valued selector and restore selected positions to their original rows before applying the existing causal action slice.
- Enable the optimization only for text-only, SP=1 models whose concrete forward signature exposes `logits_to_keep`; packed tensor selectors additionally require an explicit `torch.Tensor` annotation.
- Preserve the full-logits path for VLM/image inputs, sequence parallelism, unsupported models, and ambiguous action layouts.
- Add output and backward-parity coverage for standard and packed paths.

Fixes #1858.

## Why `num_actions + 1`

With causal shifting, the logit immediately before the first action predicts that action. Computing `A` action log-probabilities therefore requires the final `A + 1` hidden-state positions; the last selected logit is removed by the existing `[:, -A-1:-1]` slice.

## Correctness and compatibility

The optimized and full-logits paths produce matching action log-probabilities, entropy, and backward results. Packed-input coverage checks parity for every named parameter gradient; standard coverage checks the LM head and trunk. Packed models with int-only or untyped `logits_to_keep` parameters conservatively retain the full-logits path. The model is invoked exactly once, without catching and retrying failed forwards.

The primary action-log-probability result remains `[B, A]`.

For callers using `return_output=True`, auxiliary tensors are intentionally compact on optimized paths:

- standard logits: `[B, A+1, V]`
- packed logits: `[1, K, V]`, containing selected non-padding positions
- computed entropy: `[B, A+1]` after restoring packed rows

Repository policy-worker consumers use the action window and retain the same values. External callers that require full prompt-length auxiliary tensors need to account for this change.

## Benchmark

A controlled single-H200 run used `Qwen/Qwen3.5-0.8B` at revision `2fc06364715b967f1860aea9cf38778875588b17`, BF16, batch size 1, text-only input, SP=1, no packing, and no gradient checkpointing.

| Sequence / actions | Variant | LM-head tokens | Peak allocated | Peak reserved |
| --- | --- | ---: | ---: | ---: |
| 8,192 / 512 | baseline | 8,192 | 59.354 GiB | 59.840 GiB |
| 8,192 / 512 | optimized | 513 | 52.488 GiB (-11.6%) | 54.162 GiB (-9.5%) |
| 16,384 / 1,024 | baseline | 16,384 | 117.005 GiB | 118.398 GiB |
| 16,384 / 1,024 | optimized | 1,025 | 102.800 GiB (-12.1%) | 104.666 GiB (-11.6%) |

These are PyTorch CUDA allocator peaks, not full-process GPU-memory measurements. The benchmark directly wrapped the Hugging Face model and did not include FSDP sharding, entropy computation, an optimizer step, or end-to-end training. No throughput claim is made from the limited cache-cleared timing samples.

## Testing

- Focused correctness suite: 8 passed, including standard and packed output/entropy/backward parity and all-parameter gradient parity for packed inputs.
- Related CPU regression tests: 21 passed.
- H200 model-wrapper tests: 3 passed.
- Real Qwen3-VL loading and fallback parity: 2 passed.
- Single-rank FSDP2 forward, entropy, backward, gradient clipping, and optimizer-step smoke test: passed.
- Ruff, Black, Gitleaks, and `git diff --check`: passed.

Scoped pytest total: **34 passed, 0 failed, 0 skipped**. Four unrelated or two-GPU cases were intentionally deselected.

## Limitations

- The host exposed one H200, so the existing two-GPU Ulysses cases were not run. SP>1 remains on the unchanged fallback path.
- The benchmark is a single-GPU HF-model microbenchmark and does not measure FSDP, packed batches, VLMs, or end-to-end training.
- Liger Kernel was not installed in the test environment. Packed selection requires an explicit tensor-valued forward annotation, so int-only or untyped implementations fall back. Incorrect third-party annotations cannot be semantically verified without invoking the model.
- Testing covered the changed wrapper and adjacent training paths; the full repository suite was not run.

---

> [!NOTE]
> **Medium Risk**
> Touches core actor forward and training log-prob/entropy math, but gated by capability checks and broad parity tests; main residual risk is external callers expecting full-length auxiliary tensors from `return_output=True`.
> 
> **Overview**
> **Adds an optional `logits_to_keep` fast path** in `HFModelWrapper` so supported text-only actors (SP=1, no image inputs) only run the LM head on the last `num_actions + 1` positions instead of the full sequence, cutting memory and compute on long prompts.
> 
> Capability is detected via `forward` signature inspection (including PEFT base models). **Packed / padding-free batches** get a tensor index selector plus `restore_indices` so log-probs and entropy are re-padded and trimmed to the same action slice as before. **VLM, sequence parallelism, models without `logits_to_keep`, ragged `num_actions` lists, and tensor selectors on models that only accept integer suffixes** stay on the unchanged full-logits path.
> 
> **Action log-probs `[B, A]` are unchanged.** With `return_output=True`, `logits` and `entropy` may be shorter on the optimized path; in-repo RL workers already slice entropy to the action window. Tests assert `logits_to_keep` wiring, parity vs full logits (outputs, entropy, gradients), and GPU CI expectations for packed vs standard forwards.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 4c0ee00ab3df7f15ff2fbe63fccb3e04d0a9e570. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: skyrl/backends/skyrl_train/workers/model_wrapper.py [hunk 1]
```diff
@@ -33,6 +36,95 @@
...
+        as_tuple=False,
+    ).flatten()
+    if packed_selector.numel() == 0:
+        return None
+    return _ActionLogitsSelection(
+        logits_to_keep=packed_selector,
+        target_indices=packed_selector,
+        restore_indices=nnz_indices.index_select(0, packed_selector),
+        output_sequence_length=output_sequence_length,
+    )
```
[the remark this answers is not in the document]
Comment (NenoL2001) [reply skyrl/backends/skyrl_train/workers/model_wrapper.py:132]: Thanks for flagging the compatibility boundary. The supported Transformers range (5.6.1-5.8.0) already declares `logits_to_keep: int | torch.Tensor` for Llama/Qwen and uses tensors directly as advanced indices. To keep arbitrary third-party implementations safe as well, commit 4c0ee00 adds a conservative guard: packed tensor selection is enabled only when the concrete forward annotation explicitly includes `torch.Tensor`; int-only or untyped implementations retain the full-logits path. The integer suffix path is unchanged, and the wrapper still never retries a failed forward.
